New circular Seq operations: rotateRight, rotateLeft, startAt

mcallisto opened the pull request:
I would like to add to `Seq` several operations, all related to considering a `Seq` (or a subtype) a circular sequence.

If you take a look at my library repo at https://github.com/scala-tessella/ring-seq, these are all the methods that I have in mind.

But I started here adding just three, since I need to get used to the intricacies of the Scala collections.

If this is of interest, I will have many more questions on how to correctly add other methods.

mcallisto committed:
New circular Seq operations: rotateRight, rotateLeft, startAt

mcallisto committed:
Improved methods descriptions

mcallisto committed:
Failing test on rotating empty Seq

mcallisto committed:
Clearer failing test

mcallisto committed:
Failing test now ok

mcallisto committed:
Improved tests

mcallisto committed:
Correct examples in descriptions

mcallisto commented:
Any positive or negative comment about this proposal?

mcallisto committed:
Improved test on String

OndrejSpanel commented on src/main/scala/scala/collection/decorators/SeqDecorator.scala:
```diff
@@ -55,4 +55,62 @@ class SeqDecorator[C, S <: IsSeq[C]](coll: C)(implicit val seq: S) {
+   *      List(1, 2, 3, 4, 5).startAt(2) => List(3, 4, 5, 1, 2)
+   * }}}
+   */
+  def startAt[B >: seq.A, That](i: IndexO)(implicit bf: BuildFrom[C, B, That]): That =
+    rotateLeft(i)
+
```
I am not sure if this method should exists. When I have read its description, I though: but this is the same as `rotateLeft`. Once I have checked the implementation, I see it really just calls it.

mcallisto replied:
@OndrejSpanel I can definitely see your point.

The reason why, at the end, I went for the proposed duplication is that `startAt` accepts an index, while `rotateLeft` accepts a number of steps. Even if the result is the same, depending on how you see the circular sequence both can feel more natural. 

Just as an example, `startAt` plays nicely with `reflectAt` another method you can find in my more complete and documented `RingSeq` library, see https://scala-tessella.github.io/ring-seq/usage.html

som-snytt commented:
`rotate` seems pretty natural. (Also `<<` and `>>`, see `BitSetDecorators`.)

I'm not sure if it's more natural as a view, like slices, an operation on indexes.

The decisions to be taken are what to ingest here and in what form, and also whether it works better as a library.

I would agree with the previous review that `startAt` is too much, but `rotateLeft` and `rotateRight`, as a simple alias to swap slices and build a result, are handy.

I see that this repo has a lower barrier to entry.

mcallisto committed:
Deleted method rotateAt and added applyO

mcallisto commented:
Thank you @som-snytt for coming back to this issue. I deleted `startAt` and added `applyO` which I feel is another needed building block when handling circular sequences.

OndrejSpanel commented:
Why the `applyO` name? What does it mean?

mcallisto commented:
> Why the `applyO` name? What does it mean?

It stands for `applyCircular` or `applyRing`, where the `O` suffix is a _circle_ or a _ring_.
In fact it behaves like `apply`, but accepts a circular index rather than a standard one.

When looking at the issue, consider that, in perspective, these other methods could have a circular alternative:

- `sliceO`
- `indexOfSliceO`
- `lastIndexOfSliceO`
- `segmentLengthO`
- `containsSliceO`
- `slidingO`

But if you think there is a better naming, I am happy to adopt it.
